swiftide/src/integrations/treesitter/splitter.rs

```diff
@@ -109,26 +109,50 @@ impl CodeSplitter {
     /// # Returns
     ///
     /// * `Vec<String>` - A vector of code chunks as strings.
-    fn chunk_node(&self, node: Node, source: &str, mut last_end: usize) -> Vec<String> {
+    fn chunk_node(
+        &self,
+        node: Node,
+        source: &str,
+        mut last_end: usize,
+        current_chunk: Option<String>,
+    ) -> Vec<String> {
         let mut new_chunks: Vec<String> = Vec::new();
-        let mut current_chunk = String::new();
+        let mut current_chunk = current_chunk.unwrap_or(String::new());
 
         for child in node.children(&mut node.walk()) {
-            if child.end_byte() - child.start_byte() > self.max_bytes() {
-                // Child is too big, recursively chunk the child
-                if !current_chunk.is_empty() && current_chunk.len() > self.min_bytes() {
-                    new_chunks.push(current_chunk);
+            assert!(
+                current_chunk.len() <= self.max_bytes(),
+                "Chunk too big: {} > {}",
+                current_chunk.len(),
+                self.max_bytes()
+            );
+
+            // if the next child will make the chunk too big then there are two options:
+            // 1. if the next child is too big to fit in a whole chunk, then recursively chunk it one level down
+            // 2. if the next child is small enough to fit in a chunk, then add the current chunk to the list and start a new chunk
+
+            let next_child_size = child.end_byte() - last_end;
+            if current_chunk.len() + next_child_size >= self.max_bytes() {
+                // if !current_chunk.is_empty() && current_chunk.len() > self.min_bytes() {
+                //     new_chunks.push(current_chunk);
+                // }
+
+                if next_child_size > self.max_bytes() {
+                    let mut sub_chunks =
+                        self.chunk_node(child, source, last_end, Some(current_chunk));
+                    current_chunk = sub_chunks.pop().unwrap_or(String::new());
+                    new_chunks.extend(sub_chunks);
+                } else {
+                    // NOTE: if the current chunk was smaller than then the min_bytes, then it is discarded here
+                    if !current_chunk.is_empty() && current_chunk.len() > self.min_bytes() {
+                        new_chunks.push(current_chunk);
```

Comment:
This is now the only remaining location where an undersized `current_chunk` might be pushed onto `new_chunks`. If we want to eliminate this case as well I think we need to give up on this tree traversal thing and just flatten the whole thing only walking the leave nodes.

Reply:
Treesitter has bidirectional walking. text_splitter uses this in an iterator to fix it instead. It's a very rusty implementation.

Reply:
I'm not sure what you mean by that, do you mean you've got a solution in mind?

Reply:
I think it's a solvable problem by using the walker instead of a recursive for loop. But I don't think it's really that much of an easy to reimplement the whole thing right now. https://github.com/benbrandt/text-splitter/blob/main/src/splitter/code.rs#L194 is a good example. Personally I'd just let it be and get this merged 👯